File: sdk/src/wrap.ts
import type { Vendor } from "./types.js";

/**
 * wrap() internals: intercept OpenAI / Anthropic client calls and read
 * token counts from the response usage fields - streaming included.
 *
 * Structural typing only (zero runtime dependencies): a client is
 * recognized by the methods it carries, never by instanceof. Everything
 * not intercepted passes straight through a Proxy, and every observation
 * is wrapped in try/catch - an SDK bug must never break the host app.
 */

export interface RecordedCall {
  vendor: Vendor;
  model: string;
  inputTokens: number;
  outputTokens: number;
}

export type RecordCall = (call: RecordedCall) => void;

type AnyObject = Record<PropertyKey, unknown>;
type AnyFn = (...args: unknown[]) => unknown;

const isObject = (v: unknown): v is AnyObject => typeof v === "object" && v !== null;
const isFn = (v: unknown): v is AnyFn => typeof v === "function";

function safe(fn: () => void): void {
  try {
    fn();
  } catch {
    // fail open - observation errors never reach the host app
  }
}

function methodAt(obj: unknown, path: string[]): unknown {
  let cur: unknown = obj;
  for (const key of path) {
    if (!isObject(cur)) return undefined;
    cur = cur[key];
  }
  return cur;
}

export function detectVendor(client: unknown): Vendor | null {
  if (isFn(methodAt(client, ["messages", "create"]))) return "anthropic";
  if (
    isFn(methodAt(client, ["chat", "completions", "create"])) ||
    isFn(methodAt(client, ["responses", "create"])) ||
    isFn(methodAt(client, ["embeddings", "create"]))
  ) {
    return "openai";
  }
  return null;
}

/** Read token counts out of a usage object, whichever dialect it speaks:
 * OpenAI chat (prompt/completion_tokens), OpenAI responses + Anthropic
 * (input/output_tokens), embeddings (prompt_tokens only). */
function tokensFromUsage(usage: unknown): { input: number; output: number } | null {
  if (!isObject(usage)) return null;
  const num = (v: unknown) =>
    typeof v === "number" && Number.isFinite(v) && v >= 0 ? v : null;
  const input = num(usage.input_tokens) ?? num(usage.prompt_tokens);
  const output = num(usage.output_tokens) ?? num(usage.completion_tokens);
  if (input === null && output === null) return null;
  return { input: input ?? 0, output: output ?? 0 };
}

function modelOf(value: unknown): string | null {
  return isObject(value) && typeof value.model === "string" && value.model.length > 0
    ? value.model
    : null;
}

function recordFromResponse(
  vendor: Vendor,
  res: unknown,
  params: unknown,
  record: RecordCall,
): void {
  if (!isObject(res)) return;
  const tokens = tokensFromUsage(res.usage);
  const model = modelOf(res) ?? modelOf(params);
  if (!tokens || !model) return; // no usage reported = nothing to count
  record({ vendor, model, inputTokens: tokens.input, outputTokens: tokens.output });
}

interface StreamObserver {
  onChunk: (chunk: unknown) => void;
  onEnd: () => void;
}

/**
 * Accumulate usage across stream events. Counters are cumulative on both
 * vendors (Anthropic message_start carries input, message_delta the running
 * output; OpenAI's final chunk carries the whole usage), so max() per side
 * is exact. Records once, when the consumer finishes (or abandons) the
 * stream - and only if usage was actually observed: no fake numbers.
 */
function makeStreamObserver(
  vendor: Vendor,
  params: unknown,
  record: RecordCall,
): StreamObserver {
  let model = modelOf(params);
  let input: number | null = null;
  let output: number | null = null;
  let done = false;

  const onChunk = (chunk: unknown) => {
    if (!isObject(chunk)) return;
    let usage: unknown;
    if (vendor === "anthropic") {
      if (chunk.type === "message_start" && isObject(chunk.message)) {
        model = modelOf(chunk.message) ?? model;
        usage = chunk.message.usage;
      } else if (chunk.type === "message_delta") {
        usage = chunk.usage;
      }
    } else {
      model = modelOf(chunk) ?? model;
      usage = chunk.usage;
      if (chunk.type === "response.completed" && isObject(chunk.response)) {
        model = modelOf(chunk.response) ?? model;
        usage = chunk.response.usage;
      }
    }
    const tokens = tokensFromUsage(usage);
    if (tokens) {
      input = Math.max(input ?? 0, tokens.input);
      output = Math.max(output ?? 0, tokens.output);
    }
  };

  const onEnd = () => {
    if (done) return;
    done = true;
    if (model && (input !== null || output !== null)) {
      record({ vendor, model, inputTokens: input ?? 0, outputTokens: output ?? 0 });
    }
  };

  return { onChunk, onEnd };
}

/** Proxy a stream object: observe chunks through its async iterator, keep
 * every other property working against the original. */
function wrapStream<T extends object>(stream: T, observer: StreamObserver): T {
  return new Proxy(stream, {
    get(target, prop) {
      if (prop === Symbol.asyncIterator) {
        return () => {
          const inner = (target as AsyncIterable<unknown>)[Symbol.asyncIterator]();
          const iterator: AsyncIterator<unknown> = {
            async next(...args) {
              const result = await inner.next(...args);
              if (result.done) safe(observer.onEnd);
              else safe(() => observer.onChunk(result.value));
              return result;
            },
            async return(value?: unknown) {
              safe(observer.onEnd);
              return inner.return
                ? inner.return(value)
                : { done: true as const, value: undefined };
            },
            async throw(err?: unknown) {
              safe(observer.onEnd);
              if (inner.throw) return inner.throw(err);
              throw err;
            },
          };
          return iterator;
        };
      }
      const value = Reflect.get(target, prop, target);
      // Bind methods to the real object: vendor SDK classes use #private
      // fields, which break when `this` is the proxy.
      return isFn(value) ? value.bind(target) : value;
    },
  });
}

function wrapCreate(
  orig: AnyFn,
  owner: object,
  vendor: Vendor,
  isChatCompletions: boolean,
  record: RecordCall,
): AnyFn {
  return function wrappedCreate(...args: unknown[]): unknown {
    let params = args[0];
    const streaming = isObject(params) && params.stream === true;

    // OpenAI chat streams only report usage when asked: inject
    // stream_options.include_usage so streamed calls still count.
    if (streaming && vendor === "openai" && isChatCompletions && isObject(params)) {
      const streamOptions = isObject(params.stream_options) ? params.stream_options : {};
      if (streamOptions.include_usage === undefined) {
        params = { ...params, stream_options: { ...streamOptions, include_usage: true } };
        args = [params, ...args.slice(1)];
      }
    }

    const result = orig.apply(owner, args);
    if (!isObject(result) || !isFn((result as { then?: unknown }).then)) {
      return result;
    }
    const promise = result as unknown as Promise<unknown>;

    if (!streaming) {
      // Side-listen and return the ORIGINAL promise untouched, so vendor
      // promise extras (.withResponse() etc.) keep working.
      safe(() => {
        promise.then(
          (res) => safe(() => recordFromResponse(vendor, res, params, record)),
          () => {}, // the caller owns errors; nothing to count on failure
        );
      });
      return result;
    }

    const observer = makeStreamObserver(vendor, params, record);
    return promise.then((stream) =>
      isObject(stream) ? wrapStream(stream, observer) : stream,
    );
  };
}

/** Anthropic's messages.stream() returns a MessageStream; its finalMessage
 * event carries the complete usage. Attach if the shape allows, else skip. */
function wrapMessagesStream(orig: AnyFn, owner: object, record: RecordCall): AnyFn {
  return function wrappedStream(...args: unknown[]): unknown {
    const result = orig.apply(owner, args);
    safe(() => {
      const on = isObject(result) ? result.on : undefined;
      if (!isFn(on)) return;
      on.call(result, "finalMessage", (message: unknown) => {
        safe(() => {
          if (!isObject(message)) return;
          const tokens = tokensFromUsage(message.usage);
          const model = modelOf(message) ?? modelOf(args[0]);
          if (tokens && model) {
            record({
              vendor: "anthropic",
              model,
              inputTokens: tokens.input,
              outputTokens: tokens.output,
            });
          }
        });
      });
    });
    return result;
  };
}

interface Leaf {
  path: string[];
  kind: "create" | "stream";
  chat?: boolean;
}

const LEAVES: Record<Vendor, Leaf[]> = {
  anthropic: [
    { path: ["messages", "create"], kind: "create" },
    { path: ["messages", "stream"], kind: "stream" },
  ],
  openai: [
    { path: ["chat", "completions", "create"], kind: "create", chat: true },
    { path: ["responses", "create"], kind: "create" },
    { path: ["embeddings", "create"], kind: "create" },
  ],
};

/** Wrap a vendor client: known call paths are intercepted, everything else
 * passes through untouched. */
export function wrapClient<T extends object>(
  client: T,
  vendor: Vendor,
  record: RecordCall,
): T {
  const leaves = LEAVES[vendor];

  function proxyAt(target: object, prefix: string[]): object {
    return new Proxy(target, {
      get(t, prop) {
        const value = Reflect.get(t, prop, t);
        if (typeof prop !== "string") {
          return isFn(value) ? value.bind(t) : value;
        }
        const next = [...prefix, prop];
        const leaf = leaves.find(
          (l) => l.path.length === next.length && l.path.every((s, i) => s === next[i]),
        );
        if (leaf && isFn(value)) {
          return leaf.kind === "stream"
            ? wrapMessagesStream(value, t, record)
            : wrapCreate(value, t, vendor, Boolean(leaf.chat), record);
        }
        const onPath = leaves.some(
          (l) => l.path.length > next.length && next.every((s, i) => s === l.path[i]),
        );
        if (onPath && isObject(value)) return proxyAt(value, next);
        return isFn(value) ? value.bind(t) : value;
      },
    });
  }

  return proxyAt(client, []) as T;
}

File: sdk-py/src/ai_pnl/wrap.py
"""wrap() internals: intercept OpenAI / Anthropic client calls and read
token counts from the response usage fields - streaming included.

Structural typing only (zero runtime dependencies): a client is recognized
by the methods it carries, never by isinstance. Everything not intercepted
passes straight through, and every observation is wrapped in try/except -
an SDK bug must never break the host app.
"""

from __future__ import annotations

import functools
import inspect
import math
from typing import Any, Callable, NamedTuple, Optional, Tuple


class RecordedCall(NamedTuple):
    vendor: str  # "openai" | "anthropic"
    model: str
    input_tokens: int
    output_tokens: int


RecordFn = Callable[[RecordedCall], None]


def _safe(fn: Callable[[], None]) -> None:
    try:
        fn()
    except Exception:
        pass  # fail open - observation errors never reach the host app


def _field(obj: Any, name: str) -> Any:
    """Read a field structurally: attribute first (vendor SDKs return typed
    objects), mapping second (parsed JSON)."""
    if obj is None:
        return None
    try:
        value = getattr(obj, name)
        if value is not None:
            return value
    except Exception:
        pass
    try:
        if isinstance(obj, dict):
            return obj.get(name)
    except Exception:
        pass
    return None


def _method_at(obj: Any, path: Tuple[str, ...]) -> Optional[Callable[..., Any]]:
    cur = obj
    for name in path:
        if cur is None:
            return None
        try:
            cur = getattr(cur, name)
        except Exception:
            return None
    return cur if callable(cur) else None


def detect_vendor(client: Any) -> Optional[str]:
    if _method_at(client, ("messages", "create")):
        return "anthropic"
    if (
        _method_at(client, ("chat", "completions", "create"))
        or _method_at(client, ("responses", "create"))
        or _method_at(client, ("embeddings", "create"))
    ):
        return "openai"
    return None


def _count(value: Any) -> Optional[float]:
    if isinstance(value, bool):
        return None
    if isinstance(value, (int, float)) and math.isfinite(value) and value >= 0:
        return value
    return None


def _tokens_from_usage(usage: Any) -> Optional[Tuple[int, int]]:
    """Token counts out of a usage object, whichever dialect it speaks:
    OpenAI chat (prompt/completion_tokens), OpenAI responses + Anthropic
    (input/output_tokens), embeddings (prompt_tokens only)."""
    if usage is None:
        return None
    inp = _count(_field(usage, "input_tokens"))
    if inp is None:
        inp = _count(_field(usage, "prompt_tokens"))
    out = _count(_field(usage, "output_tokens"))
    if out is None:
        out = _count(_field(usage, "completion_tokens"))
    if inp is None and out is None:
        return None
    return (int(inp or 0), int(out or 0))


def _model_of(value: Any) -> Optional[str]:
    model = _field(value, "model")
    return model if isinstance(model, str) and model else None


def _record_from_response(vendor: str, res: Any, params: Any, record: RecordFn) -> None:
    tokens = _tokens_from_usage(_field(res, "usage"))
    model = _model_of(res) or _model_of(params)
    if not tokens or not model:
        return  # no usage reported = nothing to count
    record(RecordedCall(vendor, model, tokens[0], tokens[1]))


class _StreamObserver:
    """Accumulate usage across stream events. Counters are cumulative on both
    vendors (Anthropic message_start carries input, message_delta the running
    output; OpenAI's final chunk carries the whole usage), so max() per side
    is exact. Records once, when the consumer finishes (or abandons) the
    stream - and only if usage was actually observed: no fake numbers."""

    def __init__(self, vendor: str, params: Any, record: RecordFn) -> None:
        self._vendor = vendor
        self._record = record
        self._model = _model_of(params)
        self._input: Optional[int] = None
        self._output: Optional[int] = None
        self._done = False

    def on_chunk(self, chunk: Any) -> None:
        usage: Any = None
        if self._vendor == "anthropic":
            ctype = _field(chunk, "type")
            if ctype == "message_start":
                message = _field(chunk, "message")
                self._model = _model_of(message) or self._model
                usage = _field(message, "usage")
            elif ctype == "message_delta":
                usage = _field(chunk, "usage")
        else:
            self._model = _model_of(chunk) or self._model
            usage = _field(chunk, "usage")
            if _field(chunk, "type") == "response.completed":
                response = _field(chunk, "response")
                self._model = _model_of(response) or self._model
                usage = _field(response, "usage")
        tokens = _tokens_from_usage(usage)
        if tokens:
            self._input = max(self._input or 0, tokens[0])
            self._output = max(self._output or 0, tokens[1])

    def on_end(self) -> None:
        if self._done:
            return
        self._done = True
        if self._model and (self._input is not None or self._output is not None):
            self._record(
                RecordedCall(self._vendor, self._model, self._input or 0, self._output or 0)
            )


class _WrappedStream:
    """Iterates like the vendor stream it wraps, observing chunks on the way
    through; everything else delegates to the original object."""

    def __init__(self, inner: Any, observer: _StreamObserver) -> None:
        self._inner = inner
        self._observer = observer
        self._iter: Any = None

    def __iter__(self) -> "_WrappedStream":
        return self

    def __next__(self) -> Any:
        if self._iter is None:
            self._iter = iter(self._inner)
        try:
            chunk = next(self._iter)
        except BaseException:
            _safe(self._observer.on_end)  # StopIteration or a vendor error
            raise
        _safe(lambda: self._observer.on_chunk(chunk))
        return chunk

    def __enter__(self) -> "_WrappedStream":
        enter = getattr(self._inner, "__enter__", None)
        if callable(enter):
            enter()
        return self

    def __exit__(self, *exc: Any) -> Any:
        _safe(self._observer.on_end)
        exit_ = getattr(self._inner, "__exit__", None)
        return exit_(*exc) if callable(exit_) else False

    def close(self) -> None:
        _safe(self._observer.on_end)
        close = getattr(self._inner, "close", None)
        if callable(close):
            close()

    def __getattr__(self, name: str) -> Any:
        return getattr(object.__getattribute__(self, "_inner"), name)


class _WrappedAsyncStream:
    """Async twin of _WrappedStream."""

    def __init__(self, inner: Any, observer: _StreamObserver) -> None:
        self._inner = inner
        self._observer = observer
        self._iter: Any = None

    def __aiter__(self) -> "_WrappedAsyncStream":
        return self

    async def __anext__(self) -> Any:
        if self._iter is None:
            self._iter = self._inner.__aiter__()
        try:
            chunk = await self._iter.__anext__()
        except BaseException:
            _safe(self._observer.on_end)  # StopAsyncIteration or a vendor error
            raise
        _safe(lambda: self._observer.on_chunk(chunk))
        return chunk

    async def __aenter__(self) -> "_WrappedAsyncStream":
        aenter = getattr(self._inner, "__aenter__", None)
        if callable(aenter):
            await aenter()
        return self

    async def __aexit__(self, *exc: Any) -> Any:
        _safe(self._observer.on_end)
        aexit = getattr(self._inner, "__aexit__", None)
        return await aexit(*exc) if callable(aexit) else False

    async def close(self) -> None:
        _safe(self._observer.on_end)
        close = getattr(self._inner, "close", None)
        if callable(close):
            result = close()
            if inspect.isawaitable(result):
                await result

    def __getattr__(self, name: str) -> Any:
        return getattr(object.__getattribute__(self, "_inner"), name)


def _observe_result(
    result: Any, vendor: str, params: Any, streaming: bool, record: RecordFn
) -> Any:
    if inspect.isawaitable(result):
        return _observe_awaited(result, vendor, params, streaming, record)
    if streaming:
        if hasattr(result, "__iter__"):
            return _WrappedStream(result, _StreamObserver(vendor, params, record))
        return result  # not a stream shape we know - pass through untouched
    _safe(lambda: _record_from_response(vendor, result, params, record))
    return result


async def _observe_awaited(
    awaitable: Any, vendor: str, params: Any, streaming: bool, record: RecordFn
) -> Any:
    res = await awaitable  # vendor errors propagate to the caller untouched
    try:
        if streaming:
            if hasattr(res, "__aiter__"):
                return _WrappedAsyncStream(res, _StreamObserver(vendor, params, record))
            return res
        _safe(lambda: _record_from_response(vendor, res, params, record))
    except Exception:
        pass
    return res


def _wrap_create(orig: Callable[..., Any], vendor: str, is_chat: bool, record: RecordFn) -> Any:
    @functools.wraps(orig)
    def wrapped(*args: Any, **kwargs: Any) -> Any:
        params: Any = kwargs
        if args and isinstance(args[0], dict):
            params = {**args[0], **kwargs}
        streaming = params.get("stream") is True if isinstance(params, dict) else False

        # OpenAI chat streams only report usage when asked: inject
        # stream_options.include_usage so streamed calls still count.
        if streaming and vendor == "openai" and is_chat:
            try:
                opts = kwargs.get("stream_options") or {}
                if isinstance(opts, dict) and "include_usage" not in opts:
                    kwargs = {**kwargs, "stream_options": {**opts, "include_usage": True}}
            except Exception:
                pass

        result = orig(*args, **kwargs)
        try:
            return _observe_result(result, vendor, params, streaming, record)
        except Exception:
            return result

    return wrapped


class _WrappedMessageStreamManager:
    """Anthropic's messages.stream() returns a context manager whose
    MessageStream accumulates a message snapshot (usage included) no matter
    how it is consumed - events, text_stream, or get_final_message(). The
    with-block gets the REAL stream; we record from the snapshot on exit."""

    def __init__(self, manager: Any, params: Any, record: RecordFn) -> None:
        self._manager = manager
        self._params = params
        self._record = record
        self._stream: Any = None

    def __enter__(self) -> Any:
        self._stream = self._manager.__enter__()
        return self._stream

    def __exit__(self, *exc: Any) -> Any:
        _safe(self._record_snapshot)
        return self._manager.__exit__(*exc)

    def _record_snapshot(self) -> None:
        snapshot = getattr(self._stream, "current_message_snapshot", None)
        tokens = _tokens_from_usage(_field(snapshot, "usage"))
        model = _model_of(snapshot) or _model_of(self._params)
        if tokens and model:
            self._record(RecordedCall("anthropic", model, tokens[0], tokens[1]))

    def __getattr__(self, name: str) -> Any:
        return getattr(object.__getattribute__(self, "_manager"), name)


class _WrappedAsyncMessageStreamManager:
    """Async twin of _WrappedMessageStreamManager."""

    def __init__(self, manager: Any, params: Any, record: RecordFn) -> None:
        self._manager = manager
        self._params = params
        self._record = record
        self._stream: Any = None

    async def __aenter__(self) -> Any:
        self._stream = await self._manager.__aenter__()
        return self._stream

    async def __aexit__(self, *exc: Any) -> Any:
        _safe(self._record_snapshot)
        return await self._manager.__aexit__(*exc)

    def _record_snapshot(self) -> None:
        snapshot = getattr(self._stream, "current_message_snapshot", None)
        tokens = _tokens_from_usage(_field(snapshot, "usage"))
        model = _model_of(snapshot) or _model_of(self._params)
        if tokens and model:
            self._record(RecordedCall("anthropic", model, tokens[0], tokens[1]))

    def __getattr__(self, name: str) -> Any:
        return getattr(object.__getattribute__(self, "_manager"), name)


def _wrap_messages_stream(orig: Callable[..., Any], record: RecordFn) -> Any:
    @functools.wraps(orig)
    def wrapped(*args: Any, **kwargs: Any) -> Any:
        result = orig(*args, **kwargs)
        try:
            if hasattr(result, "__aenter__"):
                return _WrappedAsyncMessageStreamManager(result, kwargs, record)
            if hasattr(result, "__enter__"):
                return _WrappedMessageStreamManager(result, kwargs, record)
        except Exception:
            pass
        return result

    return wrapped


# (path, kind, is_chat_completions)
_LEAVES = {
    "anthropic": (
        (("messages", "create"), "create", False),
        (("messages", "stream"), "stream", False),
    ),
    "openai": (
        (("chat", "completions", "create"), "create", True),
        (("responses", "create"), "create", False),
        (("embeddings", "create"), "create", False),
    ),
}


class _ClientProxy:
    """Pass-through proxy: known call paths are intercepted, everything else
    reaches the real client untouched."""

    def __init__(self, target: Any, prefix: Tuple[str, ...], vendor: str, record: RecordFn) -> None:
        object.__setattr__(self, "_target", target)
        object.__setattr__(self, "_prefix", prefix)
        object.__setattr__(self, "_vendor", vendor)
        object.__setattr__(self, "_record", record)

    def __getattr__(self, name: str) -> Any:
        target = object.__getattribute__(self, "_target")
        prefix = object.__getattribute__(self, "_prefix")
        vendor = object.__getattribute__(self, "_vendor")
        record = object.__getattribute__(self, "_record")
        value = getattr(target, name)
        path = prefix + (name,)
        for leaf_path, kind, is_chat in _LEAVES[vendor]:
            if leaf_path == path and callable(value):
                if kind == "stream":
                    return _wrap_messages_stream(value, record)
                return _wrap_create(value, vendor, is_chat, record)
        on_path = any(
            len(leaf_path) > len(path) and leaf_path[: len(path)] == path
            for leaf_path, _kind, _chat in _LEAVES[vendor]
        )
        if on_path and value is not None and not callable(value):
            return _ClientProxy(value, path, vendor, record)
        return value

    def __setattr__(self, name: str, value: Any) -> None:
        setattr(object.__getattribute__(self, "_target"), name, value)

    def __repr__(self) -> str:
        return repr(object.__getattribute__(self, "_target"))

    def __dir__(self) -> Any:
        return dir(object.__getattribute__(self, "_target"))


def wrap_client(client: Any, vendor: str, record: RecordFn) -> Any:
    return _ClientProxy(client, (), vendor, record)
